libibverbs/device.c

```diff
@@ -262,6 +267,8 @@ struct ibv_context *__ibv_open_device(struct ibv_device *device)
 	context->cmd_fd = cmd_fd;
 	pthread_mutex_init(&context->mutex, NULL);
 
+	ibverbs_device_hold(device);
+
```

Comment:
That is what I was asking for originally.

Plus you have to do some solution for everything that touches sysfs. Again, maybe openat() is the answer there.

> Otherwise we can safely continue to allocate the context by calling the kernel device. In case that by that time the device was somehow dropped the API will fail and an error will be returned as expected.

I think you also need to check that the device hasn't been discovered to be removed before even attempting to open.